### ISSUE TITLE
Segmentation Fault in Eyeball filled graveyard with Lua Script.

### ISSUE THREAD

**Issue by Jastiva** [ValyriaTear/ValyriaTear#623]
I started my game and got all this, I fought some graveyard eyeballs then it crashed, I'm using the master branch committed 4 days ago.
src/valyriatear
WARNING:/home/lori/ValTear/ValyriaTear/src/common/global/objects/global_inventory_handler.cpp:GetGlobalObject:282: 
attempted to retrieve an object from inventory that didn't exist with id: 3004
ERROR:/home/lori/ValTear/ValyriaTear/src/modes/map/map_events.cpp:_Start:381: 
Error while loading ScriptedEvent start function
ERROR:/home/lori/ValTear/ValyriaTear/src/utils/src/script/script.cpp:HandleLuaError:82: 
a runtime Lua error has occured with the following error message:
  
ERROR:/home/lori/ValTear/ValyriaTear/src/utils/src/script/script.cpp:HandleLuaError:84: 
...y/layna_village/layna_village_center_at_night_script.lua:481: attempt to call method 'RemoveCharacter' (a nil value)
WARNING:/home/lori/ValTear/ValyriaTear/src/common/global/actors/global_character_handler.cpp:AddCharacter:25: 
attempted to add a character that already existed: 2
ERROR:/home/lori/ValTear/ValyriaTear/src/modes/map/map_object_supervisor.cpp:FindPath:711: 
source node coordinates are the same as the destination
ERROR:/home/lori/ValTear/ValyriaTear/src/modes/map/map_events.cpp:_Start:692: 
No path to destination (88, 79) for sprite: 3
WARNING:/home/lori/ValTear/ValyriaTear/src/modes/map/map_event_supervisor.cpp:StartEvent:98: 
The event: 'Push the rolling stone 2' is already active and can be active only once at a time. The StartEvent() call will be ignored.
 You should fix the map script: data/story/mt_elbrus/mt_elbrus_cave2_script.lua
WARNING:/home/lori/ValTear/ValyriaTear/src/modes/map/map_event_supervisor.cpp:StartEvent:98: 
The event: 'Push the rolling stone 2' is already active and can be active only once at a time. The StartEvent() call will be ignored.
 You should fix the map script: data/story/mt_elbrus/mt_elbrus_cave2_script.lua
WARNING:/home/lori/ValTear/ValyriaTear/src/modes/map/map_event_supervisor.cpp:StartEvent:98: 
The event: 'Push the rolling stone 2' is already active and can be active only once at a time. The StartEvent() call will be ignored.
 You should fix the map script: data/story/mt_elbrus/mt_elbrus_cave2_script.lua
WARNING:/home/lori/ValTear/ValyriaTear/src/common/global/objects/global_inventory_handler.cpp:GetGlobalObject:282: 
attempted to retrieve an object from inventory that didn't exist with id: 3004
WARNING:/home/lori/ValTear/ValyriaTear/src/common/global/objects/global_inventory_handler.cpp:GetGlobalObject:282: 
attempted to retrieve an object from inventory that didn't exist with id: 3004
Segmentation fault

**Comment by Bertram25** [comment]
Hey @Jastiva :)

Thanks for the bug report. I'm sorry if you encountered bad things in the master branch. I'm still in the middle of making it more stable and hopefully a bit less boring. ;)
I'll have a look ASAP.

**Comment by Bertram25** [comment]
- [ ] layna_village/layna_village_center_at_night_script.lua:481: attempt to call method 'RemoveCharacter' (a nil value) Fix access to CharacterHandler.
- [ ] attempted to retrieve an object from inventory that didn't exist with id: 3004 (Probably because of going to the item in the skill graph)
- [ ] Segfault in the cemetery.
Not bad in themselves, but would be nice to be handled better:
- [ ] ERROR:/home/lori/ValTear/ValyriaTear/src/modes/map/map_object_supervisor.cpp:FindPath:711:
source node coordinates are the same as the destination (Should not be a warning)
- [ ] The event: 'Push the rolling stone 2' is already active and can be active only once at a time. The StartEvent() call will be ignored. (Should not be a warning in default mode.) --> Should become a true IF_PRINT_WARNING(debug mode).

**Comment by Bertram25** [comment]
```
Thread 1 "valyriatear" received signal SIGSEGV, Segmentation fault.
0x0000555555d2d24c in vt_audio::AudioDescriptor::FadeIn (this=0x5555589ede50, time=1000) at /home/yohann/Projets/ValyriaTear/src/engine/audio/audio_descriptor.cpp:625
625             Play();
(gdb) bt
#0  0x0000555555d2d24c in vt_audio::AudioDescriptor::FadeIn (this=0x5555589ede50, time=1000) at /home/yohann/Projets/ValyriaTear/src/engine/audio/audio_descriptor.cpp:625
#1  0x0000555555eaff40 in vt_map::private_map::SoundObject::ApplyVolume (this=0x5555589dfc40) at /home/yohann/Projets/ValyriaTear/src/modes/map/map_objects/map_sound.cpp:153
#2  0x0000555555e9c34e in vt_map::private_map::ObjectSupervisor::_UpdateAmbientSounds (this=0x555558d78ed0) at /home/yohann/Projets/ValyriaTear/src/modes/map/map_object_supervisor.cpp:420
#3  0x0000555555e9b9bd in vt_map::private_map::ObjectSupervisor::Update (this=0x555558d78ed0) at /home/yohann/Projets/ValyriaTear/src/modes/map/map_object_supervisor.cpp:275
#4  0x0000555555e8b49b in vt_map::MapMode::Update (this=0x5555586199e0) at /home/yohann/Projets/ValyriaTear/src/modes/map/map_mode.cpp:361
#5  0x0000555555d3a116 in vt_mode_manager::ModeEngine::Update (this=0x555556341700) at /home/yohann/Projets/ValyriaTear/src/engine/mode_manager.cpp:333
#6  0x000055555601dc21 in main (argc=1, argv=0x7fffffffe098) at /home/yohann/Projets/ValyriaTear/src/main.cpp:609
```
Found this, triggered in the cemetery.

And this, whenever you use BattleItems:
```
WARNING:/home/yohann/Projets/ValyriaTear/src/modes/battle/battle_item.cpp:~BattleItem:30: 
actual count was not equal to available count upon destruction
```

**Comment by Bertram25** [comment]
@Jastiva I would like to express a special thanks to you for this bug report as it made me realize there was quite a flaw in the sound management code.

For those interested (and for own later reference), the segmentation fault was triggered by Audio Descriptor pointers becoming flawed in the MapMode sound object list.
This was due to the fact that the Audio Manager had an audio cache that limited the active audio descriptors to 16. It means that when the 17th sound was attempted to be played or loaded, the first one not playing in the active descriptor would get dropped.
For some reason, (certainly the right number of audio sounds being loaded to flaw the map object one), this happened a lot more in the Cemetery story map and helped me a lot narrow down that one.

Funnily, the amount of audio source limit is 64 while the permitted audio descriptors were 16, so the loaded audio source were never fully used.

So for the moment, I simply dropped the audio cache loading limit and made it so the first audio source with a fully unloaded descriptor (not just stopped), or a never loaded one can be used as an audio source.
This is not perfect as if someone ever (and I bet it might be me) load more than 64 sounds at once, some sounds won't be able to play at all.
I thus wonder then why the audio source addition/removal is not part of the audio descriptor management, so that you always have the same amount of source than active audio descriptor and that's all. I've opened this issue to think about investigating this improvement later on: #624
